File: cocktail-shaker-sort/cocktail-shaker-sort.go
package main

import (
	"math/rand"
	"time"
	"fmt"
)

func random_sequence(minimum, maximum int) []int { //returns a shuffled array
	array := make([]int, 0)
	for i := minimum; i < maximum; i++ { //create array and append numbers to it
		array = append(array, i)
	}

	rand.Seed(time.Now().UnixNano())
	rand.Shuffle(len(array), func(i, j int) { array[i], array[j] = array[j], array[i] }) //shuffle the array

	return array
}

func cocktail_shaker_sort(array []int) []int {
	var swaps int //number of swaps
	var section int
	for amt := 0; amt < len(array)/2 + len(array) % 2; amt++ { //iterate for half the length of the array plus one if odd
		swaps = 0

		if amt > 0 {
			section = len(array[amt:len(array)-amt])
		} else {
			section = len(array)
		}

		for i := 0; i < section-1; i++ { //iterate through entire array
			if array[amt+i] > array[amt+i+1] { //swap if needed
                array[amt+i], array[amt+i+1] = array[amt+i+1], array[amt+i]
                swaps ++
			}
		}

		if swaps == 0 { //array is in order, this break statement could save time
			break
		}

		for i := section-1; i > 0; i-- {
			if array[amt+i] < array[amt+i-1] { //swap if needed
                array[amt+i], array[amt+i-1] = array[amt+i-1], array[amt+i]
                swaps ++
			}
		}

		if swaps == 0 { //array is in order, this break statement could save time
			break
		}
	}
	return array
}

func isSorted(array []int) bool {
	for i := 0; i < len(array)-1; i++ {
		if array[i] > array[i+1] {
			return false
		}
	}
	return true
}

func main() {
	shuffled_array := random_sequence(0, 1000)
	fmt.Println("COCKTAIL SHAKER SORT")
	fmt.Println(shuffled_array)
	fmt.Println()
	fmt.Println(cocktail_shaker_sort(shuffled_array))
}

File: cocktail-shaker-sort/cocktail-shaker-sort.py
def cocktail_shaker_sort(array): #COCKTAIL SHAKER SORT
    for amt in range(len(array)//2 + len(array)%2): #iterate for half the length of the array plus one if odd
        swaps = 0
        section = len(array[amt:-amt]) if amt > 0 else len(array) #keep narrowing down the search

        for i in range(section-1): #iterate through the array
            if array[amt+i] > array[amt+i+1]: #swap if needed
                array[amt+i], array[amt+i+1] = array[amt+i+1], array[amt+i]
                swaps += 1

        if swaps == 0: break #array is in order, this break statement could save time

        for i in range(section-1, 0, -1):
            if array[amt+i] < array[amt+i-1]: #swap if needed
                array[amt+i], array[amt+i-1] = array[amt+i-1], array[amt+i]
                swaps += 1

        if swaps == 0: break #array is in order

    return array


if __name__ == "__main__":
    import os, sys #import shuffler from parent directory
    dir_path = os.path.dirname(os.path.realpath(__file__))
    parent_dir_path = os.path.abspath(os.path.join(dir_path, os.pardir))
    sys.path.insert(0, parent_dir_path)
    from shuffler import random_sequence

    print("COCKTAIL SHAKER SORT")

    shuffled_array = random_sequence(0, 1000)

    print(shuffled_array, "\n")

    print(cocktail_shaker_sort(shuffled_array))